### PR TITLE
Update Cycle Calendar (CYCLE_DATES)

### PR DESCRIPTION
**Pull request by TobyRet**
## Context

A cycle should start with Find opening and end with Find closing

## Changes proposed in this pull request

- Update `CYCLE_DATES` with Find opening and end with Find closing
- Add 2022 cycle dates
- Updated the CycleTimetableHelper to play nicely with specs as  a result of calendar changes (returns static dates)
- Fix all the broken things

## Guidance to review

Recommend review by commit as these are quite fundamental changes. 

## Link to Trello card

https://trello.com/c/narHEhQx/3544-%F0%9F%94%81-eoc-change-calendar-to-start-with-find-opening-and-end-with-find-closing

## Things to check

- [x] This code does not rely on migrations in the same Pull Request
- [ ] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [ ] API release notes have been updated if necessary
- [ ] Required environment variables have been updated [added to the Azure KeyVault](/docs/environment-variables.md#deploy-pipeline)

### CODE REVIEW

File: spec/support/test_helpers/cycle_timetable_helper.rb [hunk 1]
```diff
@@ -1,39 +1,17 @@
 module CycleTimetableHelper
   def mid_cycle
-    previous_end_of_cycle_timetable[:apply_reopens] + 1.day
+    CycleTimetable::CYCLE_DATES.dig(2020, :apply_opens) + 1.day
```
Comment (davidgisbey) [inline spec/support/test_helpers/cycle_timetable_helper.rb:3]: So the point of this was to randomise the times that tests run at to catch edge cases.

 I don't think it's a huge deal but would it have worked as expected if the dates were set to the current and next cycle, rather than previous and current i think it would have travelled forward in time rather than back?

File: app/services/cycle_timetable.rb [hunk 2]
```diff
@@ -33,10 +35,14 @@ def self.current_year
     CYCLE_DATES.keys.detect do |year|
       return year if year == CYCLE_DATES.keys.last
```
Comment (davidgisbey) [inline app/services/cycle_timetable.rb:36]: So this returns the latest year when the next cycle launches right?

i.e. if it's after find reopens it will return 2022?

Sorry if i'm being dense, but how does this differ from just doing

```
return CYCLE_DATES.keys.last
```

Comment (davidgisbey) [reply app/services/cycle_timetable.rb:36]: Have discussed. Me being stupid. Makes sense to me! TIL about detect

File: spec/support/test_helpers/cycle_timetable_helper.rb (hunk 1 above)
Comment (TobyRet) [reply spec/support/test_helpers/cycle_timetable_helper.rb:3]: I believe as a result of the calendar changes, we need to work with a cycles that has a complete set of dates. We do not have all the dates for 2022 so this means we need to work with 2020 & 2021 data (or earlier)

Comment (TobyRet) [reply spec/support/test_helpers/cycle_timetable_helper.rb:3]: I can understand your point about randomising times to catch edge cases. I think it this strategy has been successful at catching edge cases it may be work re-instating. But on the other hand, having a constant date makes debugging a bit less painful (but that's just me!)

File: app/services/cycle_timetable.rb (hunk 2 above)
Comment (davidgisbey) [reply app/services/cycle_timetable.rb:36]: I wonder if this could be something like 

```
return year if return_year_if_next_cycles_dates_not_set

private

def return_year_if_next_cycles_dates_not_set
  year == CYCLE_DATES.keys.last
end 
```

Just to convey intent

File: spec/support/test_helpers/cycle_timetable_helper.rb (hunk 1 above)
Comment (davidgisbey) [reply spec/support/test_helpers/cycle_timetable_helper.rb:3]: Yea i totally get that. I wonder if we can just add the next few years dates with approximations then update when required? Ideally going forward we can standardise each date i.e. find opening on the first tues of Oct but that seems a way off.

File: app/services/cycle_timetable.rb [hunk 3]
```diff
@@ -62,29 +68,29 @@ def self.find_closes
...
 
   def self.between_cycles_apply_2?
     Time.zone.now > date(:apply_2_deadline).end_of_day &&
-      Time.zone.now < date(:apply_reopens).beginning_of_day
+      Time.zone.now < date(:apply_opens, next_year).beginning_of_day
   end
 
-  def self.date(name)
-    schedule = schedules.fetch(current_cycle_schedule)
+  def self.date(name, year = current_year)
```
Comment (duncanjbrown) [inline app/services/cycle_timetable.rb:92]: :+1: This is a nice way to keep the `_reopens` methods

File: app/services/cycle_timetable.rb [hunk 4]
```diff
@@ -33,10 +35,14 @@ def self.current_year
     CYCLE_DATES.keys.detect do |year|
       return year if year == CYCLE_DATES.keys.last
 
-      now.between?(CYCLE_DATES[year][:find_reopens], CYCLE_DATES[year + 1][:find_reopens])
+      now.between?(CYCLE_DATES[year][:find_opens], CYCLE_DATES[year + 1][:find_opens])
```
Comment (duncanjbrown) [inline app/services/cycle_timetable.rb:38]:
```suggestion
      now.between?(date(:find_opens, year), date(:find_opens, year + 1))
```

File: app/services/cycle_timetable.rb [hunk 5]
```diff
@@ -162,7 +168,8 @@ def self.current_cycle?(application_form)
   end
 
   def self.can_add_course_choice?(application_form)
-    return true if Time.zone.now.to_date >= find_reopens && !application_form.must_be_carried_over?
+    return false if application_form.must_be_carried_over?
+    return true if Time.zone.now.to_date >= find_reopens
```
Comment (duncanjbrown) [inline app/services/cycle_timetable.rb:172]: Will this ever be true? `find_reopens` is now by definition a date in the _next_ cycle

File: app/services/cycle_timetable.rb [hunk 6]
```diff
@@ -180,7 +187,7 @@ def self.before_find_reopens?
   end
 
   def self.before_apply_reopens?
-    return true if Time.zone.now.to_date <= apply_reopens.beginning_of_day
+    return true if Time.zone.now.to_date <= CYCLE_DATES[next_year][:apply_opens].beginning_of_day
```
Comment (duncanjbrown) [inline app/services/cycle_timetable.rb:190]:
```suggestion
    return true if Time.zone.now.to_date <= apply_reopens.beginning_of_day
```

What should this method return if we're already into the new cycle? Do we need to consider `:apply_opens` in this cycle too?

File: app/views/candidate_interface/carry_over/start_between_cycles_unsubmitted.html.erb [hunk 7]
```diff
@@ -12,7 +12,7 @@
 
     <% if CycleTimetable.before_apply_reopens? %>
```
Comment (duncanjbrown) [inline app/views/candidate_interface/carry_over/start_between_cycles_unsubmitted.html.erb:13]: See above — this is inconsistent because it's possible to be `before_apply_reopens?` before we've entered the new cycle

File: spec/services/cycle_timetable_spec.rb [hunk 8]
```diff
@@ -91,19 +105,19 @@
 
   describe '.find_down?' do
     it 'returns false before find closes' do
-      Timecop.travel(CycleTimetable.find_closes.beginning_of_day - 1.hour) do
+      Timecop.travel(CycleTimetable::CYCLE_DATES[2020][:find_closes].beginning_of_day - 1.hour) do
```
Comment (duncanjbrown) [inline spec/services/cycle_timetable_spec.rb:108]: nit: it'd be good to use the method interface rather than reaching down into the hash here. Maybe by adding an optional `year` param to `find_closes`?

**Review by duncanjbrown — COMMENTED** [review]
This is a massive improvement. Couple of qs inline about `apply_opens` vs `apply_reopens`

File: app/services/cycle_timetable.rb (hunk 2 above)
Comment (TobyRet) [reply app/services/cycle_timetable.rb:36]: Yeah that's a good point. I think there are a few cases in this class where we could extract to method. 
All the methods in this class are `class methods` so we will need to declare private class methods (`private_class_method : next_cycles_dates_not_set?`)

File: app/services/cycle_timetable.rb (hunk 4 above)
Comment (TobyRet) [reply app/services/cycle_timetable.rb:38]: Ah yes good spot 👍

File: spec/support/test_helpers/cycle_timetable_helper.rb (hunk 1 above)
Comment (TobyRet) [reply spec/support/test_helpers/cycle_timetable_helper.rb:3]: Hey, I'm easy about adding approximations for dates in the future. But worth considering up until this point we have approximated dates for previous years only to make some of the tests pass. I think this may be a bit of a code smell really. I think Duncan had an idea regarding creating a calendar especially for testing purposes which would resolve this however.

File: app/services/cycle_timetable.rb (hunk 5 above)
Comment (TobyRet) [reply app/services/cycle_timetable.rb:172]: @duncanjbrown @davidgisbey - we commented out line 172 (`find_reopens`) and the entire test suite has still passed.  I'm not entirely sure what the implications are regarding business rules. Can you please advise?

File: app/services/cycle_timetable.rb (hunk 6 above)
Comment (TobyRet) [reply app/services/cycle_timetable.rb]: `.before_apply_reopens?` is called from the carry over interstitials (and nowhere else currently) - this needs to compare the current time versus when apply reopens and we do not need to consider when apply opened (yet!)

Comment (TobyRet) [reply app/services/cycle_timetable.rb]: https://github.com/DFE-Digital/apply-for-teacher-training/pull/5106/commits/e34524361f6910fef5dcb5eb999fa32c2d40da5e

File: app/services/cycle_timetable.rb (hunk 5 above)
Comment (duncanjbrown) [reply app/services/cycle_timetable.rb:176]: Good excuse to add a spec for this method!

File: spec/services/cycle_timetable_spec.rb (hunk 8 above)
Comment (TobyRet) [reply spec/services/cycle_timetable_spec.rb]: Yep good idea. Done

Comment (TobyRet) [reply spec/services/cycle_timetable_spec.rb]: https://github.com/DFE-Digital/apply-for-teacher-training/pull/5106/commits/eb018b3009f02d29bb6efc86dfd6994cd89b8e1c

File: app/services/cycle_timetable.rb (hunk 5 above)
Comment (TobyRet) [reply app/services/cycle_timetable.rb:180]: Actually there is a spec it passes because another one of the condition is true. You are right Duncan - it is now impossible for line 172  to ever resolve to 'true' as it now always looks ahead. I'm wondering if this is has implications or if that line can simply be removed?

File: app/views/candidate_interface/carry_over/start_between_cycles_unsubmitted.html.erb (hunk 7 above)
Comment (TobyRet) [reply app/views/candidate_interface/carry_over/start_between_cycles_unsubmitted.html.erb:13]: These carry over interstitials are only displayed if Time.now is `between_cycles_apply_2` - see https://github.com/DFE-Digital/apply-for-teacher-training/blob/master/app/controllers/candidate_interface/carry_over_controller.rb#L7

I think perhaps some more tweaks / refactoring may be required when we come to updating the interstitials (which are separate cards)

File: app/services/cycle_timetable.rb (hunk 5 above)
Comment (TobyRet) [reply app/services/cycle_timetable.rb:180]: Interestingly, there are specs for this (e.g. https://github.com/DFE-Digital/apply-for-teacher-training/pull/5106/files#diff-fad1017966158907e50ba7ac1d694af33bc81613fb3d7771a667d78d806615c4R168) however you are 100% right Duncan. Line 172 will now always resolve to 'true' as we are now always looking ahead to when Find reopens. The reason why the specs pass is because of another condition below resolving to true.

I'm not sure if for now we can simply delete that line or if this has implications that we have not thought of.

File: spec/components/candidate_interface/deadline_banner_component_spec.rb [hunk 9]
```diff
@@ -18,15 +18,15 @@ def configure_conditions_for_rendering_banner(phase)
 
       result = render_inline(described_class.new(phase: application_form.phase, flash_empty: flash.empty?))
 
-      expect(result.text).to include('The deadline for applying to courses starting this academic year is 24 August')
+      expect(result.text).to include("The deadline for applying to courses starting this academic year is #{CycleTimetable.apply_1_deadline.strftime('%d %B')}")
```
Comment (stevehook) [inline spec/components/candidate_interface/deadline_banner_component_spec.rb:21]: Maybe worth adding `'%d %B'` to `Date::DATE_FORMATS`.

File: app/services/cycle_timetable.rb [hunk 10]
```diff
@@ -95,64 +109,64 @@ def self.current_cycle_schedule
...
         apply_1_deadline: 4.days.ago.to_date,
         apply_2_deadline: 2.days.ago.to_date,
         find_closes: 1.day.ago.to_date,
-        find_reopens: 1.day.from_now.to_date,
-        apply_reopens: 2.days.from_now.to_date,
+        find_opens: 1.day.from_now.to_date,
+        apply_opens: 2.days.from_now.to_date,
       },
 
       today_is_after_find_reopens: {
```
Comment (stevehook) [inline app/services/cycle_timetable.rb:156]: Is this name is a little confusing now? Should it just be `today_is_after_find_opens`?

File: app/services/cycle_timetable.rb [hunk 11]
```diff
@@ -95,64 +109,64 @@ def self.current_cycle_schedule
...
         apply_1_deadline: 5.days.ago.to_date,
         apply_2_deadline: 3.days.ago.to_date,
         find_closes: 2.days.ago.to_date,
-        find_reopens: 1.day.ago.to_date,
-        apply_reopens: 1.day.from_now.to_date,
+        find_opens: 1.day.ago.to_date,
+        apply_opens: 1.day.from_now.to_date,
       },
 
       today_is_after_apply_reopens: {
```
Comment (stevehook) [inline app/services/cycle_timetable.rb:164]: Should this be `today_is_after_apply_opens`?

**Review by stevehook — COMMENTED** [review]
This looks great. I've made a couple of very minor suggestions.

File: app/lib/test_applications.rb [hunk 12]
```diff
@@ -537,7 +537,7 @@ def initialize_time(recruitment_cycle_year)
       earliest_date = 20.days.ago.to_date
       latest_date = Time.zone.now.to_date
     else
-      earliest_date = CycleTimetable::CYCLE_DATES[recruitment_cycle_year][:apply_reopens]
+      earliest_date = CycleTimetable::CYCLE_DATES[recruitment_cycle_year][:apply_opens]
```
Comment (duncanjbrown) [inline app/lib/test_applications.rb:540]: this could use the new param to `apply_opens`? from eb018b3

File: app/models/performance_statistics.rb [hunk 13]
```diff
@@ -202,12 +202,12 @@ def application_choices
   end
 
   def date_range_query_for_recruitment_cycle_year(cycle_year)
-    start_date = CycleTimetable::CYCLE_DATES[cycle_year][:apply_reopens]
+    start_date = CycleTimetable::CYCLE_DATES[cycle_year][:apply_opens]
```
Comment (duncanjbrown) [inline app/models/performance_statistics.rb:205]:
```suggestion
    start_date = CycleTimetable::CYCLE_DATES.apply_opens(cycle_year)
```

etc

File: spec/components/candidate_interface/offer_review_component_spec.rb [hunk 14]
```diff
@@ -13,7 +13,7 @@
            course_option: course_option,
            application_form: application_form)
   end
-  let(:find_closes) { CycleTimetable::CYCLE_DATES.dig(Time.zone.now.year, :find_closes) }
+  let(:find_closes) { CycleTimetable::CYCLE_DATES.dig(RecruitmentCycle.previous_year, :find_closes) }
```
Comment (duncanjbrown) [inline spec/components/candidate_interface/offer_review_component_spec.rb:16]:
```suggestion
  let(:find_closes) { CycleTimetable::CYCLE_DATES.find_closes(RecruitmentCycle.previous_year) }
```

File: app/services/cycle_timetable.rb (hunk 5 above)
Comment (TobyRet) [reply app/services/cycle_timetable.rb]: David has suggested some tweaks to the '.can_add_course_choice?' which deals with the issues above and IMO greatly improves readability. Thanks @davidgisbey 
https://github.com/DFE-Digital/apply-for-teacher-training/pull/5106/commits/01d5d2774b0343481e491b5efdafa9df51cd1ced

File: spec/components/candidate_interface/offer_review_component_spec.rb [hunk 15]
```diff
@@ -4,7 +4,7 @@
   include CourseOptionHelpers
 
   let(:course_option) { course_option_for_provider_code(provider_code: 'ABC') }
-  let(:application_form) { create(:application_form, submitted_at: Time.zone.now) }
+  let(:application_form) { create(:application_form, submitted_at: Time.zone.now - 1.year) }
```
Comment (duncanjbrown) [inline spec/components/candidate_interface/offer_review_component_spec.rb:7]: We don't need this? specs pass without

File: app/views/support_interface/settings/cycles.html.erb [hunk 16]
```diff
@@ -38,7 +38,7 @@
   'Apply 2 deadline' => CycleTimetable.apply_2_deadline.to_s(:govuk_date),
   'Find closes on' => CycleTimetable.find_closes.to_s(:govuk_date),
   'Find reopens on' => CycleTimetable.find_reopens.to_s(:govuk_date),
-  'Apply reopens on' => CycleTimetable.apply_reopens.to_s(:govuk_date),
+  'Apply reopens on' => CycleTimetable.apply_opens.to_s(:govuk_date),
```
Comment (duncanjbrown) [inline app/views/support_interface/settings/cycles.html.erb:41]: This should be `.reopens`

**Review by duncanjbrown — APPROVED** [review]
2 last comments! and 🚀 🚀 🚀 🚀 🚀 🚀

File: spec/components/candidate_interface/offer_review_component_spec.rb (hunk 15 above)
Comment (TobyRet) [reply spec/components/candidate_interface/offer_review_component_spec.rb]: Correct! Removed 😸

File: app/views/support_interface/settings/cycles.html.erb (hunk 16 above)
Comment (TobyRet) [reply app/views/support_interface/settings/cycles.html.erb]: Yes!
https://github.com/DFE-Digital/apply-for-teacher-training/pull/5106/commits/14ecdf560c1cb5b42c301c526ef381748cf26e9a
